File: oneflow/core/autograd/gradient_funcs/where.cpp
/*
Copyright 2020 The OneFlow Authors. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
#include "oneflow/core/framework/op_expr_grad_function.h"
#include "oneflow/core/functional/functional.h"
#include "oneflow/core/functional/impl/common.h"

namespace oneflow {
namespace one {

struct WhereCaptureState : public AutoGradCaptureState {
  bool requires_grad_x = false;
  bool requires_grad_y = false;
  DimVector x_reduce_dims = {};
  DimVector y_reduce_dims = {};
  DimVector x_squeeze_dims = {};
  DimVector y_squeeze_dims = {};
};

class Where : public OpExprGradFunction<WhereCaptureState> {
 public:
  Maybe<void> Init(const OpExpr& op) override;
  Maybe<void> Capture(WhereCaptureState* ctx, const TensorTuple& inputs, const TensorTuple& outputs,
                      const AttrMap& attrs) const override;
  Maybe<void> Apply(const WhereCaptureState* ctx, const TensorTuple& out_grads,
                    TensorTuple* in_grads) const override;
};

Maybe<void> Where::Init(const OpExpr& op) { return Maybe<void>::Ok(); }

Maybe<void> Where::Capture(WhereCaptureState* ctx, const TensorTuple& inputs,
                           const TensorTuple& outputs, const AttrMap& attrs) const {
  // cond, x, y
  CHECK_EQ_OR_RETURN(inputs.size(), 3);  // NOLINT(maybe-need-error-msg)
  ctx->requires_grad_x = inputs.at(1)->requires_grad();
  ctx->requires_grad_y = inputs.at(2)->requires_grad();
  if ((!ctx->requires_grad_x) && (!ctx->requires_grad_y)) { return Maybe<void>::Ok(); }

  ctx->SaveTensorForBackward(inputs.at(0));  // condition

  CHECK_EQ_OR_RETURN(outputs.size(), 1);
  const Shape& out_shape = *outputs.at(0)->shape();
  auto GetReduceDims = [&](DimVector& reduce_dim_vec, DimVector& squeeze_dim_vec,
                           const std::shared_ptr<oneflow::one::Tensor>& tensor) -> Maybe<void> {
    reduce_dim_vec.clear();
    squeeze_dim_vec.clear();
    const Shape& shape = *tensor->shape();
    if (functional::IsScalarTensor(tensor)) {
      reduce_dim_vec.resize(out_shape.size());
      squeeze_dim_vec.resize(out_shape.size());
      std::iota(reduce_dim_vec.begin(), reduce_dim_vec.end(), 0);
      std::iota(squeeze_dim_vec.begin(), squeeze_dim_vec.end(), 0);
    } else if (shape != out_shape) {
      CHECK_GE_OR_RETURN(out_shape.size(), shape.size());  // NOLINT(maybe-need-error-msg)
      size_t ddiff = out_shape.size() - shape.size();
      for (int i = 0; i < out_shape.size(); ++i) {
        if (i < ddiff) {
          reduce_dim_vec.push_back(i);
          squeeze_dim_vec.push_back(i);
        } else if (out_shape[i] != shape[i - ddiff]) {
          reduce_dim_vec.push_back(i);
        }
      }
    }
    return Maybe<void>::Ok();
  };
  JUST(GetReduceDims(ctx->x_reduce_dims, ctx->x_squeeze_dims, inputs.at(1)));
  JUST(GetReduceDims(ctx->y_reduce_dims, ctx->y_squeeze_dims, inputs.at(2)));

  return Maybe<void>::Ok();
}

Maybe<void> Where::Apply(const WhereCaptureState* ctx, const TensorTuple& out_grads,
                         TensorTuple* in_grads) const {
  if ((!ctx->requires_grad_x) && (!ctx->requires_grad_y)) { return Maybe<void>::Ok(); }
  CHECK_EQ_OR_RETURN(out_grads.size(), 1);  // NOLINT(maybe-need-error-msg)
  const auto& out_grad = out_grads.at(0);
  CHECK_EQ_OR_RETURN(ctx->SavedTensors().size(), 1);  // NOLINT(maybe-need-error-msg)
  const auto& condition = ctx->SavedTensors().at(0);
  std::shared_ptr<oneflow::one::Tensor> zero;
  if (out_grad->is_local()) {
    zero = JUST(
        functional::Constant(Shape({}), Scalar(0), out_grad->dtype(), JUST(out_grad->device())));
  } else {
    const size_t sbp_ndim = JUST(out_grad->nd_sbp())->sbp_parallel_size();
    std::vector<Symbol<SbpParallel>> nd_sbp_vec;
    nd_sbp_vec.reserve(sbp_ndim);
    for (int i = 0; i < sbp_ndim; ++i) {
      SbpParallel sbp;
      sbp.mutable_broadcast_parallel();
      nd_sbp_vec.push_back(SymbolOf(sbp));
    }
    const auto& parallel_desc = JUST(out_grad->parallel_desc());
    zero = JUST(functional::GlobalConstant(Shape({}), Scalar(0), out_grad->dtype(), parallel_desc,
                                           nd_sbp_vec));
  }
  in_grads->resize(3);  // cond, x, y
  if (ctx->requires_grad_x) {
    auto x_grad = JUST(functional::Where(condition, out_grad, zero));
    if (!ctx->x_reduce_dims.empty()) {
      x_grad = JUST(functional::ReduceSum(
          x_grad, std::vector<int32_t>{ctx->x_reduce_dims.begin(), ctx->x_reduce_dims.end()},
          /*keepdims=*/true));
    }
    if (!ctx->x_squeeze_dims.empty()) {
      x_grad = JUST(functional::Squeeze(
          x_grad, std::vector<int32_t>{ctx->x_squeeze_dims.begin(), ctx->x_squeeze_dims.end()}));
    }
    in_grads->at(1) = x_grad;
  }
  if (ctx->requires_grad_y) {
    auto y_grad = JUST(functional::Where(condition, zero, out_grad));
    if (!ctx->y_reduce_dims.empty()) {
      y_grad = JUST(functional::ReduceSum(
          y_grad, std::vector<int32_t>{ctx->y_reduce_dims.begin(), ctx->y_reduce_dims.end()},
          /*keepdims=*/true));
    }
    if (!ctx->y_squeeze_dims.empty()) {
      y_grad = JUST(functional::Squeeze(
          y_grad, std::vector<int32_t>{ctx->y_squeeze_dims.begin(), ctx->y_squeeze_dims.end()}));
    }
    in_grads->at(2) = y_grad;
  }
  return Maybe<void>::Ok();
}

REGISTER_OP_EXPR_GRAD_FUNCTION("where", Where);

}  // namespace one
}  // namespace oneflow

File: python/oneflow/nn/modules/where.py
"""
Copyright 2020 The OneFlow Authors. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""
import oneflow as flow
from oneflow.framework.tensor import register_tensor_op


@register_tensor_op("where")
def where_op(condition, x, y):
    """Return a tensor of elements selected from either :attr:`x` or :attr:`y`, depending on :attr:`condition`.
    If the element in condition is larger than 0,

    it will take the `x` element, else it will take the `y` element

    .. note::

        The tensors :attr:`condition`, :attr:`x`, :attr:`y` must be broadcastable.
        It will take the `x` element, else it will take the `y` element.

    Args:
        condition (IntTensor): When 1 (nonzero), yield x, otherwise yield y
        x (Tensor or Scalar): value (if :attr:x is a scalar) or values selected at indices
                            where :attr:`condition` is True
        y (Tensor or Scalar): value (if :attr:x is a scalar) or values selected at indices
                            where :attr:`condition` is False
    Returns:
        Tensor: A tensor of shape equal to the broadcasted shape of :attr:`condition`, :attr:`x`, :attr:`y`

    For example:

    .. code-block:: python

        >>> import numpy as np
        >>> import oneflow as flow
        >>> x = flow.Tensor(
        ...    np.array([[-0.4620, 0.3139], [0.3898, -0.7197], [0.0478, -0.1657]]),
        ...    dtype=flow.float32,
        ... )
        >>> y = flow.Tensor(np.ones(shape=(3, 2)), dtype=flow.float32)
        >>> condition = flow.Tensor(np.array([[0, 1], [1, 0], [1, 0]]), dtype=flow.int32)
        >>> out = condition.where(x, y)
        >>> out #doctest: +ELLIPSIS
        tensor([[1.    , 0.3139],
                ...
                [0.0478, 1.    ]], dtype=oneflow.float32)

    """

    assert condition.dtype == flow.int32 or condition.dtype == flow.int8
    if isinstance(x, int) or isinstance(x, float):
        x = flow.Tensor(
            [float(x)], dtype=flow.float32, device=flow.device(condition.device.type),
        )
    if isinstance(y, int) or isinstance(y, float):
        y = flow.Tensor(
            [float(y)], dtype=flow.float32, device=flow.device(condition.device.type),
        )
    assert (
        condition.device.type == x.device.type
        and condition.device.type == y.device.type
    )
    assert len(condition.shape) == len(x.shape) and len(condition.shape) == len(
        y.shape
    ), f"The dim of where module's inputs can not match, please check!"
    return flow.F.where(condition, x, y)


if __name__ == "__main__":
    import doctest

    doctest.testmod(raise_on_error=True)
